src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+                {
+                    ensureSpace(cur, end, literal.size(), "required literal size not matched");
+                    if (std::string_view(cur, literal.size()) != literal)
+                        throw Exception(
+                            ErrorCodes::LOGICAL_ERROR, "Expect literal {} but {} provided", literal, std::string_view(cur, literal.size()));
+                    cur += literal.size();
+                    return cur;
+                }
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber2(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 2, "readNumber2 requires size >= 2");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber3(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 3, "readNumber4 requires size >= 3");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber4(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 4, "readNumber4 requires size >= 4");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            static void ensureSpace(Pos cur, Pos end, size_t len, const String & msg)
+            {
+                if (cur > end || cur + len > end)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unable to parse because {}", msg);
+            }
+
+            template <bool check_space = true>
+            static Pos assertChar(Pos cur, Pos end, char ch)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 1, "assertChar requires size >= 1");
+
+                if (*cur != ch)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Expect char {}, but {} provided", String(ch, 1), String(*cur, 1));
+
+                ++cur;
+                return cur;
+            }
+
+            static Pos mysqlDayOfWeekTextShort(Pos cur, Pos end, Date & date)
+            {
+                ensureSpace(cur, end, 3, "Parsing DayOfWeekTextShort requires size >= 3");
```

Comment:
Also, the user provided a MySQL/joda format string and won't know internal function names. So instead of throwing an error with message `DayOfWeekTextShort`, the error should include something like `%a`.

Reply:
You perhaps want to include the entire parsed string in the error message. Imagine function `parseDateTime()` runs on millions of rows. If there is an error on l. 746284901, it will be impossible for the user to debug when he/she doesn't get to see the misformed data.

Reply:
> Minor: In case of an error, the message will have slightly odd lower/uppercase, e.g. `Unable to parse because Parsing MonthOfYearTextShort requires size >= 3`. The messages passed in from outside would need to start with a lowercase character, e.g. l. 734.

fixed

Reply:
> You perhaps want to include the entire parsed string in the error message. Imagine function `parseDateTime()` runs on millions of rows. If there is an error on l. 746284901, it will be impossible for the user to debug when he/she doesn't get to see the misformed data.

I'm worrying the introducing of the parsed string may slow down performance because of temporarily constructed/destructed strings.

Reply:
Anyway, I already added flag and parsed string in almost every exception...